### PR TITLE
Account page

### PR DESCRIPTION
**Pull request by Brianna-Flo**
## Description

<what this pull request is doing>
Allow a user to store their preferences in an account page
Allow a user to edit their email and preferences after account creation
Fetch user data from database
Implement constants and reuse components

<what will be done in later PRs and not included here>
Add error state component (currently on a separate unmerged branch)
Filter recipes based on user preferences

## Milestones
<the milestones or stories/features that this works towards>
Frontend design
Integrating firebase authentication into database to store user information

## Resources
<links to tutorials, code snippets, inspirations>
<if you took or translated specific code from the internet, please call it out here!>
Updating a users email: https://firebase.google.com/docs/auth/web/manage-users
Authenticating a user: https://firebase.google.com/docs/reference/kotlin/com/google/firebase/auth/EmailAuthCredential
Enums: https://maxheiber.medium.com/alternatives-to-typescript-enums-50e4c16600b1
https://www.internalfb.com/wiki/ReScript/Enums/

## Test Plan
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>
<div>
    <a href="https://www.loom.com/share/da2fa2f0aa6e4900934d94ec79a51bcd">
      <p>Edit Account Details Demo - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/da2fa2f0aa6e4900934d94ec79a51bcd">
      
    </a>
  </div>

### CODE REVIEW

File: frontend/src/pages/AccountPage.tsx [hunk 1]
```diff
@@ -7,9 +7,35 @@ import { useState } from "react";
...
+import { auth } from "../utils/firebase"
+import { onAuthStateChanged } from "firebase/auth";
+import type { User } from "firebase/auth";
+
 const AccountPage = ({ navOpen, toggleNav }: RecipeToggleNavBar) => {
   const [userIntollerances, setUserIntollerances] = useState<string[]>([]);
   const [userDiets, setUserDiets] = useState<string[]>([]);
+  const [currentUser, setCurrentUser] = useState<User | null>()
+
+//   const auth = getAuth();
```
Comment (nithyad) [inline frontend/src/pages/AccountPage.tsx]: delete

File: frontend/src/pages/AccountPage.tsx [hunk 2]
```diff
@@ -7,9 +7,35 @@ import { useState } from "react";
...
+    const checkUserSignedIn = () => {
+        onAuthStateChanged(auth, (user) => {
+          if (user) {
+            // User is signed in, see docs for a list of available properties
+            // https://firebase.google.com/docs/reference/js/auth.user
+            const userSignedIn = auth.currentUser
+            setCurrentUser(userSignedIn)
+            console.log("current user is", userSignedIn);
+          //   console.log("user", userProfile?.email, "is signed in")
```
Comment (nithyad) [inline frontend/src/pages/AccountPage.tsx]: delete commented out statement

File: frontend/src/pages/AccountPage.tsx [hunk 3]
```diff
@@ -7,9 +7,35 @@ import { useState } from "react";
...
+            // ...
+          } else {
+            // User is signed out
+            // ...
+            console.log("there is no signed in user");
```
Comment (nithyad) [inline frontend/src/pages/AccountPage.tsx]: what is the plan with this- instead of just logging there is no signed in user are you going to have an error pop up or something if the user isnt logged in?

File: frontend/src/components/AppHeader.tsx [hunk 4]
```diff
@@ -4,15 +4,38 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
+import { signOut } from "firebase/auth";
+import { useNavigate } from "react-router";
+import "../styles/Homepage.css"
+
 const AppHeader = ({navOpen, toggleNav}: RecipeToggleNavBar) => {
+  const navigate = useNavigate();
+
+  const handleLogout = () => {
+    signOut(auth).catch((error) => {
+      console.log("TODO: Implement error state component")
```
Comment (radhikadhomse) [inline frontend/src/components/AppHeader.tsx:17]: TODOs can just be comments rather than console logs.

File: frontend/src/pages/Homepage.tsx [hunk 5]
```diff
@@ -1,28 +1,28 @@
 import "../styles/Homepage.css";
 import type { RecipeToggleNavBar } from "../utils/types";
-import { useNavigate } from "react-router";
+// import { useNavigate } from "react-router";
```
Comment (nithyad) [inline frontend/src/pages/Homepage.tsx:3]: delete

File: frontend/src/pages/Homepage.tsx [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
 import { groceryList, ingredients } from "../utils/sampleData";
 import NextRecipe from "../components/NextRecipe";
 import IngredientsPreview from "../components/IngredientsPreview";
 import GroceryPreview from "../components/GroceryPreview";
 import AppHeader from "../components/AppHeader";
 
 const Homepage = ({ navOpen, toggleNav }: RecipeToggleNavBar) => {
-  const navigate = useNavigate();
+  // const navigate = useNavigate();
```
Comment (nithyad) [inline frontend/src/pages/Homepage.tsx:11]: delete this whole part

File: frontend/src/pages/Homepage.tsx [hunk 7]
```diff
@@ -1,28 +1,28 @@
...
+  //     toggleNav();
+  //   }
+  //   navigate("/login");
+  // };
 
   return (
     <div className="homepage-container">
       <AppHeader navOpen={navOpen} toggleNav={toggleNav} />
-      <button className="login-button" onClick={handleLoginClick}>
+      {/* <button className="login-button" onClick={handleLoginClick}>
```
Comment (nithyad) [inline frontend/src/pages/Homepage.tsx:23]: also this

File: frontend/src/styles/Homepage.css [hunk 8]
```diff
@@ -18,23 +18,31 @@
 }
 
 .nav-button {
-    position: absolute;
+    /* position: absolute; */
```
Comment (nithyad) [inline frontend/src/styles/Homepage.css:21]: delete

File: frontend/src/components/AppHeader.tsx [hunk 9]
```diff
@@ -4,15 +4,38 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
+      console.log("error during signout ;-;")
```
Comment (radhikadhomse) [inline frontend/src/components/AppHeader.tsx:18]: let's remove all the console logs when youre done debugging

File: frontend/src/styles/Homepage.css [hunk 10]
```diff
@@ -18,23 +18,31 @@
...
+    display: flex;
+    flex-direction: column;
+    margin: 1vw;
 }
 
 .login-button {
-    position: absolute;
-    right: 1vw;
-    top: 5vh;
+    /* position: absolute; */
```
Comment (nithyad) [inline frontend/src/styles/Homepage.css:37]: delete commented out part 37-39

File: frontend/src/components/AppHeader.tsx [hunk 11]
```diff
@@ -4,15 +4,38 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
+
   return (
     <section className="app-header">
-      <header>
-        <h1>Grocery Buddy *make a better name lol*</h1>
-      </header>
       <button className="nav-button" onClick={toggleNav}>
         <FontAwesomeIcon icon={faBars} className="nav-icon" />
       </button>
+      <h1>Grocery Buddy *make a better name lol*</h1>
```
Comment (radhikadhomse) [inline frontend/src/components/AppHeader.tsx:34]: let's add the *make a better name lol* as a TODO comment

File: frontend/src/components/AppHeader.tsx [hunk 12]
```diff
@@ -4,15 +4,38 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
+      <div className="auth-access-buttons">
+        <button className="login-button" onClick={handleLogout}>Logout</button>
+        <button className="login-button" onClick={handleLoginClick}>Login</button>
```
Comment (radhikadhomse) [inline frontend/src/components/AppHeader.tsx:37]: for consistency, they should both be handleX or handleXClick

File: frontend/src/pages/AccountPage.tsx [hunk 13]
```diff
@@ -1,60 +1,136 @@
...
+        prev.filter((item) => item !== value)
       );
     } else {
-      setUserIntollerances((prev) => [...prev, selectedIntollerance]);
+      setterFunction((prev) => [...prev, value]);
+    }
+  }
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = event.target;
```
Comment (radhikadhomse) [inline frontend/src/pages/AccountPage.tsx:60]: let's stop calling it name, value and make it more specifically what it is

File: frontend/src/pages/AccountPage.tsx [hunk 14]
```diff
@@ -1,60 +1,136 @@
...
-  const handleIntolleranceClick = (
-    event: React.MouseEvent<HTMLButtonElement>
-  ) => {
-    const selectedIntollerance = event.currentTarget.value;
-    if (userIntollerances.includes(selectedIntollerance)) {
-      setUserIntollerances((prev) =>
-        prev.filter((intollerance) => intollerance !== selectedIntollerance)
+  const handlePreferenceClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    const { name, value } = event.currentTarget
+    const setterFunction = name === PreferenceList.INTOLERANCES ? setUserIntolerances : setUserDiets;
```
Comment (radhikadhomse) [inline frontend/src/pages/AccountPage.tsx:48]: instead of name/value, lets use what it is
also can we cahnge PreferenceList to PreferenceListType or PreferenceListCAtegory or smthg?

File: frontend/src/components/LoginForm.tsx [hunk 15]
```diff
@@ -1,26 +1,27 @@
...
+import { Authentication } from '../utils/constants'
 
 
 
-const LoginForm = ({handleSubmit, handleChange, formData}: RecipeAuthFormEvents) => {
+const LoginForm = ({handleSubmit, handleChange, formData}: RecipeLoginFormEvents) => {
     return (
         <form className="login-info" onSubmit={handleSubmit}>
-            <label htmlFor="email">Email</label>
+            <label htmlFor={Authentication.EMAIL}>Email</label>
```
Comment (radhikadhomse) [inline frontend/src/components/LoginForm.tsx:11]: can we rename to AuthenticationFieldType.EMAIL so it's clear without checking that its an enum

File: frontend/src/pages/AccountPage.tsx (hunk 14 above)
Comment (radhikadhomse) [reply frontend/src/pages/AccountPage.tsx:48]: similarly setterFunction ➡️  setPreferenceList

File: frontend/src/pages/Homepage.tsx (hunk 5 above)
Comment (radhikadhomse) [inline frontend/src/pages/Homepage.tsx:3]: delete

File: frontend/src/pages/Homepage.tsx [hunk 16]
```diff
@@ -1,28 +1,28 @@
...
-      toggleNav();
-    }
-    navigate("/login");
-  };
+  // const handleLoginClick = () => {
+  //   if (navOpen) {
+  //     toggleNav();
+  //   }
+  //   navigate("/login");
+  // };
```
Comment (radhikadhomse) [inline frontend/src/pages/Homepage.tsx:18]: delete

File: frontend/src/pages/Homepage.tsx [hunk 17]
```diff
@@ -1,28 +1,28 @@
...
 
   return (
     <div className="homepage-container">
       <AppHeader navOpen={navOpen} toggleNav={toggleNav} />
-      <button className="login-button" onClick={handleLoginClick}>
+      {/* <button className="login-button" onClick={handleLoginClick}>
         Login
-      </button>
+      </button> */}
```
Comment (radhikadhomse) [inline frontend/src/pages/Homepage.tsx:25]: deete

File: frontend/src/pages/SignupForm.tsx [hunk 18]
```diff
@@ -2,28 +2,32 @@ import { useState } from "react";
...
 
 const SignupForm = ({ navOpen, toggleNav }: RecipeToggleNavBar) => {
   const [formData, setFormData] = useState<RecipeAuthFormData>({email: "", password: ""});
   const [message, setMessage] = useState<RecipeAuthFormResult>();
   const navigate = useNavigate();
   const [success, setSuccess] = useState(false);
 
-  const [userIntollerances, setUserIntollerances] = useState<string[]>([])
-  const [userDiets, setUserDiets] = useState<string[]>([])
+  // const [userIntolerances, setUserIntolerances] = useState<string[]>([])
```
Comment (radhikadhomse) [inline frontend/src/pages/SignupForm.tsx:18]: delete

File: frontend/src/pages/SignupForm.tsx [hunk 19]
```diff
@@ -36,9 +40,11 @@ const SignupForm = ({ navOpen, toggleNav }: RecipeToggleNavBar) => {
         const user = userCredential.user;
         setMessage({ type: "success", text: "Registration successful!" });
         setSuccess(true);
-        const newUser: RecipeNewUserFirebaseId = {
+        const newUser: RecipeUserAccountInfo = {
           firebaseId: user.uid ? user.uid : "",
           email: user.email ? user.email : "",
```
Comment (radhikadhomse) [inline frontend/src/pages/SignupForm.tsx:45]: check what happens when these are actualyl empty string and handle that case

File: frontend/src/styles/Homepage.css (hunk 8 above)
Comment (radhikadhomse) [inline frontend/src/styles/Homepage.css:21]: dleete

File: frontend/src/utils/constants.ts [hunk 20]
```diff
@@ -0,0 +1,11 @@
+const PreferenceList = {
+  INTOLERANCES: "INTOLERANCES",
+  DIETS: "DIETS"
+}
+
+const Authentication = {
+    EMAIL: "email",
+    PASSWORD: "password"
+}
+
```
Comment (radhikadhomse) [inline frontend/src/utils/constants.ts:10]: since theyre enums lets make that clear in the naming otherwise i would be expecting a full object that contains entities regarding preference list

File: frontend/src/utils/databaseHelpers.ts [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+        },
+        body: JSON.stringify({
+          email: userEmail,
+          intolerances: userIntolerances,
+          diets: userDiets,
+        }),
+      }
+    );
+    if (!updatedUser.ok) {
+      throw new Error("Failed to update user");
```
Comment (radhikadhomse) [inline frontend/src/utils/databaseHelpers.ts:20]: it will crash if we throw it

File: frontend/src/utils/databaseHelpers.ts [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+          `http://localhost:3000/account/${user.uid}`,
+          {
+            method: "GET",
+            headers: {
+              "Content-Type": "application/json",
+            },
+          }
+        );
+        if (!fetchedUserData.ok) {
+          throw new Error("Failed to get user data");
```
Comment (radhikadhomse) [inline frontend/src/utils/databaseHelpers.ts:39]: if we throw the error, it will crash
we would want to present this gracefully to the user like Sorry, we could not process your request. Please try again later
or smthg like that

File: frontend/src/utils/types.ts [hunk 23]
```diff
@@ -28,10 +32,29 @@ interface RecipeIngredientData {
     expirationDate?: string
 }
 
-interface RecipeAuthFormEvents {
+// interface RecipeAuthFormEvents {
```
Comment (radhikadhomse) [inline frontend/src/utils/types.ts:35]: omit

File: frontend/src/styles/Homepage.css [hunk 24]
```diff
@@ -18,23 +18,31 @@
...
+    /* right: 1vw; */
+    /* top: 5vh; */
```
Comment (radhikadhomse) [inline frontend/src/styles/Homepage.css:39]: delete
